### PR TITLE
Make netty 4.0 attributes definition compatible with Apache Atlas 1.1.0

### PR DESCRIPTION
**Pull request by labbati**
The issue we are trying to fix here seems to be a result of two concurrent issues.

Cassandra driver's used by atlas 1.1.0 brings in a old version of netty < 4.0.23.

In version 4.0.25 netty raised exception in `UniqueName` class for duplicate name

```
if (map.putIfAbsent(name, Boolean.TRUE) != null) {
            throw new IllegalArgumentException(String.format("'%s' is already in use", name));
        }
```

This changed in 4.0.26 where exception was not thrown, instead the duplicate key was just ignored.

```
    protected UniqueName(String name) {
        this.name = checkNotNull(name, "name");
        id = nextId.incrementAndGet();
    }
```

Now this alone would not be enough. The only way I was able to reproduce is if two different class loaders tries to load the
same class `AttributeKeys` twice.
I tried to look into Atlas source code (which is huge) and I could not find a clear path of how this would be possible.

Further more, looking at some build scripts that a user experiencing this issue provided,
it looks like they are doing everything correctly and there is nothing they should change.

In conclusion, I believe that even if this is not a very clean approach,
we should be resilient to this use case, when a customer using netty <4.0.25 in an
older version of Apache Atlas (or with some sort of class loader behavior).

### CODE REVIEW

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/ClassLoaderScopedWeakMap.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
+package datadog.trace.agent.tooling;
+
+import datadog.trace.bootstrap.WeakMap;
+import java.util.HashMap;
+import java.util.Map;
+
+public class ClassLoaderScopedWeakMap {
```
Comment (tylerbenson) [inline dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/ClassLoaderScopedWeakMap.java]: Instead of creating this separate class, can I suggest adding `getOrCreate` to `WeakMap` instead?

Comment (labbati) [reply dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/ClassLoaderScopedWeakMap.java]: To be honest I think it makes sense to have it into its own class as it introduces this concept of per-ClassLoader map that I think is slightly out of context in `WeakMap`. If you think, though, that it is better to have it in `WeakMap` after reading this comment, I will happily move it there :)

Comment (tylerbenson) [reply dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/ClassLoaderScopedWeakMap.java]: I don't feel that strongly about it either way, so use your best judgement.

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/WeakMapSuppliers.java [hunk 2]
```diff
@@ -175,9 +175,18 @@ public void put(final K key, final V value) {
       public void putIfAbsent(final K key, final V value) {
         map.putIfAbsent(key, value);
       }
+
+      @Override
+      public V getOrCreate(K key, ValueSupplier<V> supplier) {
+        if (!map.containsKey(key)) {
+          map.put(key, supplier.get());
```
Comment (tylerbenson) [inline dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/WeakMapSuppliers.java]: Consider adding "double check locking" here to ensure thread safety.

File: dd-trace-ot/src/main/java/datadog/trace/common/writer/ListWriter.java [hunk 3]
```diff
@@ -38,7 +38,7 @@ public void waitForTraces(final int number) throws InterruptedException, Timeout
       }
       latches.add(latch);
     }
-    if (!latch.await(20, TimeUnit.SECONDS)) {
+    if (!latch.await(5, TimeUnit.SECONDS)) {
```
Comment (tylerbenson) [inline dd-trace-ot/src/main/java/datadog/trace/common/writer/ListWriter.java]: Intentional?

File: dd-java-agent/instrumentation/netty-4.0/src/main/java/datadog/trace/instrumentation/netty40/client/HttpClientRequestTracingHandler.java [hunk 4]
```diff
@@ -20,6 +20,7 @@
 
   @Override
   public void write(final ChannelHandlerContext ctx, final Object msg, final ChannelPromise prm) {
+    System.out.println("[STD LOG] About to write request handler");
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/netty-4.0/src/main/java/datadog/trace/instrumentation/netty40/client/HttpClientRequestTracingHandler.java]: Remove.

File: dd-java-agent/instrumentation/netty-4.0/src/main/java/datadog/trace/instrumentation/netty40/AttributeKeys.java [hunk 5]
```diff
@@ -1,19 +1,53 @@
...
-      new AttributeKey<>(HttpClientTracingHandler.class.getName() + ".span");
+      attributeKey(HttpClientTracingHandler.class.getName() + ".span");
+
+  /**
+   * Generate an attribute key or reuse the one existing in the global app map. This implementation
+   * creates attributes only once even if the current class is loaded by several class loaders and
+   * prevents an issue with Apache Atlas project were this class loaded by multiple class loaders,
+   * while the Attribute class is loaded by a third class loader and used internally for the
+   * cassandra driver.
+   */
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/netty-4.0/src/main/java/datadog/trace/instrumentation/netty40/AttributeKeys.java]: Do we want/need a similar change in the netty 4.0 instrumentation?

File: dd-java-agent/agent-bootstrap/src/test/groovy/datadog/trace/bootstrap/WeakMapTest.groovy [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+import spock.lang.Specification
+
+class WeakMapTest extends Specification {
+
+
+  def "getOrCreate a value"() {
+    setup:
+    def supplier = new CounterSupplier()
+    def sut = new WeakMap.MapAdapter<String, Integer>(new WeakHashMap<>())
```
Comment (tylerbenson) [inline dd-java-agent/agent-bootstrap/src/test/groovy/datadog/trace/bootstrap/WeakMapTest.groovy]: These could be moved to class fields instead of having to duplicate for each test.  (Unless marked as `@Shared`, Spock will reinitialize before each test.)

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/WeakMap.java [hunk 7]
```diff
@@ -96,6 +106,15 @@ public void putIfAbsent(final K key, final V value) {
       }
     }
 
+    @Override
+    public synchronized V getOrCreate(K key, ValueSupplier<V> supplier) {
+      if (!map.containsKey(key)) {
+        map.put(key, supplier.get());
```
Comment (tylerbenson) [inline dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/WeakMap.java]: Trade `synchronized` for double check locking here instead.

File: dd-java-agent/instrumentation/netty-4.0/src/main/java/datadog/trace/instrumentation/netty40/AttributeKeys.java (hunk 5 above)
Comment (labbati) [reply dd-java-agent/instrumentation/netty-4.0/src/main/java/datadog/trace/instrumentation/netty40/AttributeKeys.java]: I think you are referring to `4.1`. I think this issue is very specific to the usage that Apache Atlas does of the class loaders + netty. We can do that as a safety net though, as I do not see any disadvantage, so good call.

**Comment by tylerbenson** [comment]
Looks like there's a conflict with master and the tests are failing.  Once those are fixed, good to merge.

**Comment by labbati** [comment]
Merge conflict resolved and added new style definition for new property in `AttributeKeys`.
